### ISSUE TITLE
torch.cat results does not have `requires_grad` if under an autograd function

### ISSUE THREAD

**Issue by wanchaol** [pytorch/pytorch#55438]
If we call torch.cat under an autograd function, its result does not have `requires_grad` set as we wanted

Example code:

```python
import torch
from torch.autograd import Function


class SimpleCatGrad(Function):
    @staticmethod
    def forward(ctx, inputs):
        input_tensor = torch.cat(inputs, dim=1).view([-1])
        print("input tensor requries-grad: " + str(input_tensor.requires_grad))
        output = input_tensor.new_empty([170])
        return output + input_tensor

    @staticmethod
    def backward(ctx, *grad_outputs):
        gard = grad_outputs[0]
        return (None, None, None, gard)


input0 = torch.randn(10, 8, requires_grad=True)
input1 = torch.randn(10, 9, requires_grad=True)

input_list = [input0, input1]

print("simple cat requires-grad: " + str(torch.cat(input_list, dim=1).view([-1]).requires_grad))

ret = SimpleCatGrad.apply(input_list)
```

output:
```
simple cat requires-grad: True
input tensor requries-grad: False
```

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer

**Comment by SsnL** [comment]
hmm this is expected? things run in `Function::forward` are without autograd graph tracking.

**Comment by albanD** [comment]
This is expected indeed, the forward of the custom Function is not tracking any gradient (running in no_grad mode). It doesn't need to because you provide the backward :)
